src/DataFrameIntervals.jl

```diff
@@ -3,44 +3,50 @@ module DataFrameIntervals
 using Intervals, DataFrames, Requires, Dates
 export quantile_windows, interval_join, groupby_interval_join
 
+# represents a lazy application of fn.(vec1, vec2, ...), useful in cases where
+# fn is cheap and we don't mind recomputing it for each call to `getindex`
+struct FnVector{T,F,A} <: AbstractVector{T}
+    fn::F
+    args::A
+end
+Base.size(x::FnVector) = size(x.args[1])
+function lazy(fn::Base.Callable, args::AbstractVector...)
+    sizes = size.(args)
+    all(==(first(sizes)), sizes) ||
+        throw(ArgumentError("Vectors must all have the same size."))
+    isempty(args) && throw(ArgumentError("Must use non-empty vectors"))
+    T = typeof(fn(getindex.(args, 1)...))
```

Comment:
Shouldn't this be something like `T = promote_type(typeof.([fn[i] for i in 1:first(size(fn))))`, or use something like BangBang.jl to widen types?

Reply:
Discussed over huddle: we should catch the error that happens here for type unstable columns and type unstable functions, and provide some convenient message.

Reply:
From pair-review, this is fine, but it'll throw an error if the function is type-unstable.

Maybe wrap methods that trigger this lazy evaluation so that those errors are caught and a user-friendly error is presented saying why things went wrong.